lib/src/common/cache/cache.dart

```diff
@@ -294,6 +301,66 @@ class Cache<TIdentifier, TValue> extends Object with Disposable {
+  /// [TIdentifier] pair from the [Cache], [get] or [getAsync] will. In other
+  /// words, it's a get that behaves like `Map`'s does rather than the way
+  /// [Cache]'s does. This can be convenient, for example, when mutating an
+  /// object that is known to be in the cache.
+  ///
+  /// This ideally shouldn't be used to create a reference with a lifetime that
+  /// will exceed the time the item exists in the cache. For example this would be
+  /// a bad consumption pattern:
+  ///
+  /// The delineation to the right of the code indicates when an item will be in
+  /// the cache and when a reference created with [weakGet] would be alive.
+  ///
+  ///     var cache = new Cache<String, Disposable>(new LeastRecentlyUsedStrategy(0));
+  ///     cache.didRemove.listen((CacheContext context) => context.value.dispose());
+  ///     cache.get('thing', () => new Disposable());                                  ┓
+  ///                                                                                  ┃
+  ///     var danglingReference = await cache.weakGet('thing');                        ┃┓
+  ///                                                                                  ┃┃
+  ///     await cache.release('thing');                                                ┛┃
+  ///     // Await a random future to ensure that the callback registered on didRemove  ┃
+  ///     // is called.                                                                 ┃
+  ///     await new Future((){});                                                       ┃
+  ///                                                                                   ┃
+  ///     // This will throw because the instance associated with 'thing' has been      ┃
+  ///     // removed from the cache and disposed on the didRemove callback.             ┃
+  ///     danglingReference.manageDisposable(new Disposable());                         ┋
+  ///
+  /// A usual consumption pattern would look something like:
+  ///
+  ///     class Things {
+  ///       Cache<String, Thing> _cache =
+  ///           new Cache<String, Thing>(new LeastRecentlyUsedStrategy(10));
+  ///
+  ///       Future<Null> createAThing(String id) async {
+  ///         await _cache.get(id, heavyThingCreation);
+  ///       }
+  ///
+  ///       /// It is expected that consumers of [Things] will call [createAThing], before
+  ///       /// [mutateAThing].
+  ///       Future<Null> mutateAThing(String id) async {
+  ///         // Using a weak get here rather than a full get will prevent unintentional
+  ///         // retention in the case where mutateAThing is called after removeAThing
+  ///         (await _cache.weakGet(id)).mutate();
+  ///       }
+  ///
+  ///       void removeAThing(String id) {
+  ///         _cache.release(id);
+  ///       }
+  ///     }
+  Future<TValue> weakGet(TIdentifier id) {
+    _throwWhenDisposed('weakGet');
+    _log.finest('weakGet id: $id');
+    return _cache[id];
```

Comment:
Yes this is intended to work like a `Map` get rather than a `Cache` get so if the value isn't there you'll get a `null` value. I can write a test for that.

Reply:
Everything is stored in the cache as a `Future<TValue>` so I'm not sure what we can do to get around the future

Reply:
Oh, I see, then it should probably be `return _cache[id] ?? Future.null()`.